qiskit/transpiler/target.py

```diff
@@ -0,0 +1,651 @@
+        ]
+        return instruction_properties[index]
+
+    def _build_coupling_graph(self):
+        self._coupling_graph = rx.PyDiGraph(multigraph=False)
+        self._coupling_graph.add_nodes_from(list({} for _ in range(self.num_qubits)))
+        for gate, qarg_map in self._gate_map.items():
+            for qarg, properties in qarg_map.items():
+                if len(qarg) == 1:
+                    self._coupling_graph[qarg[0]] = properties
+                elif len(qarg) == 2:
+                    try:
+                        edge_data = self._coupling_graph.get_edge_data(*qarg)
+                        edge_data[gate] = properties
+                    except rx.NoEdgeBetweenNodes:
+                        self._coupling_graph.add_edge(*qarg, {gate: properties})
+
+    def coupling_map(self, two_q_gate=None):
+        """Get a :class:`~qiskit.transpiler.CouplingMap` from this target.
+
+        Args:
+            two_q_gate (str): An optional gate name for a two qubit gate in
+                the Target to generate the coupling map for. If specified the
+                output coupling map will only have edges between qubits where
+                this gate is present.
+        Returns:
+            CouplingMap: The :class:`~qiskit.transpiler.CouplingMap` object
+                for this target.
+
+        Raises:
+            ValueError: If a non-two qubit gate is passed in for ``two_q_gate``.
+            IndexError: If an Instruction not in the Target is passed in for
+                ``two_q_gate``.
+        """
+        if None in self._qarg_gate_map:
+            return None
+        if any(len(x) > 2 for x in self.qargs):
+            logger.warning(
+                "This Target object contains multiqubit gates that "
+                "operate on > 2 qubits. This will not be reflected in "
+                "the output coupling map."
+            )
+
+        if two_q_gate is not None:
+            coupling_graph = rx.PyDiGraph(multigraph=False)
+            coupling_graph.add_nodes_from(list(None for _ in range(self.num_qubits)))
+            for qargs, properties in self._gate_map[two_q_gate].items():
+                if len(qargs) != 2:
+                    raise ValueError(
+                        "Specified two_q_gate: %s is not a 2 qubit instruction" % two_q_gate
+                    )
+                coupling_graph.add_edge(*qargs, {two_q_gate: properties})
+            cmap = CouplingMap()
+            cmap.graph = coupling_graph
+            return cmap
+
+        if self._coupling_graph is None:
+            self._build_coupling_graph()
+        cmap = CouplingMap()
+        cmap.graph = self._coupling_graph
+        return cmap
+
+    @property
+    def physical_qubits(self):
+        """Returns a sorted list of physical_qubits"""
+        return list(range(self.num_qubits))
+
+    def __iter__(self):
+        return iter(self._gate_map)
+
+    def __getitem__(self, key):
```

Comment:
We cannot `backend.target["sx"][(0,)]` replace with `backend.target["sx"][0]` (`qarg` must be a tuple of integers), right? How about adding something (type hint or doctring) here to tell the `qarg` in the `key` must be tuple of integers?

Reply:
Yeah, that's probably an important thing to do. The only place I can really document this is in the top level class documentation, anything I put here for `__getitem__()` won't get rendered in the documentation. Or are you suggesting I convert a raw int to a tuple as part `__getitem__()` to simplify 1q access patterns? I can do both what do you think?

Reply:
can we allow [0] instead of always tuple? Since single-qubit gates will be a majority of gates

Reply:
We can't easily because it's the inner dict key. To support this I'd have to create a custom container class to replace that inner dict which wraps a dict and would convert `[0]` to `[(0,)]` on get item call. It's a lot of boilerplate to support that

Reply:
ok then let's leave it